File: 0125_valid_palindrome.py
# https://leetcode-cn.com/problems/valid-palindrome/

class Solution:
    def isPalindrome(self, s):
        """
        Inputs:
            s: str
        Outputs:
            res: bool
        """
        
        i = 0
        j = len(s) - 1
        res = True
        while i < j:
            while (i < len(s)) and (not s[i].isalnum()):
                i = i + 1
            while (j > -1) and (not s[j].isalnum()):
                j = j - 1
            
            if i == len(s):
                break
            if s[i].lower() != s[j].lower():
                res = False
                break
            else:
                i = i + 1
                j = j - 1
        return res

File: Cpp/0125_valid_palindrome.cpp
// https://leetcode.cn/problems/valid-palindrome/

class Solution {
public:
    bool isPalindrome(string s) {
        int n = s.size();
        int i = 0, j = n - 1;
        while (i < j) {
            while ((i < n) && (!isalnum(s[i]))) {
                i = i + 1;
            }
            while ((j >= 0) && (!isalnum(s[j]))) {
                j = j - 1;
            }

            if ((i >= n) || (j < 0)) {
                continue;
            }

            if (tolower(s[i]) == tolower(s[j])) {
                i = i + 1;
                j = j - 1;
            }
            else {
                return false;
            }
        }
        return true;
    }
};
